GVFS/GVFS.Common/Database/GVFSDatabase.cs

```diff
@@ -0,0 +1,188 @@
+        private BlockingCollection<IDbConnection> connectionPool;
+
+        public GVFSDatabase(PhysicalFileSystem fileSystem, string enlistmentRoot, IDbConnectionFactory connectionFactory, int initialPooledConnections = InitialPooledConnections)
+        {
+            this.connectionPool = new BlockingCollection<IDbConnection>();
+            this.databasePath = Path.Combine(enlistmentRoot, GVFSPlatform.Instance.Constants.DotGVFSRoot, GVFSConstants.DotGVFS.Databases.VFSForGit);
+            this.connectionFactory = connectionFactory;
+
+            string folderPath = Path.GetDirectoryName(this.databasePath);
+            fileSystem.CreateDirectory(folderPath);
+
+            try
+            {
+                for (int i = 0; i < initialPooledConnections; i++)
+                {
+                    this.connectionPool.Add(this.connectionFactory.OpenNewConnection(this.databasePath));
+                }
+
+                this.Initialize();
+            }
+            catch (Exception ex)
+            {
+                throw new GVFSDatabaseException($"{nameof(GVFSDatabase)} constructor threw exception setting up connection pool and initializing", ex);
+            }
+        }
+
+        public void Dispose()
+        {
+            if (this.disposed)
+            {
+                return;
+            }
+
+            this.disposed = true;
+            this.connectionPool.CompleteAdding();
+            while (!this.connectionPool.IsCompleted && this.connectionPool.TryTake(out IDbConnection connection))
+            {
+                connection.Dispose();
+            }
+
+            this.connectionPool.Dispose();
+            this.connectionPool = null;
+        }
+
+        IDbConnection IGVFSConnectionPool.GetConnection()
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(GVFSDatabase));
+            }
+
+            IDbConnection connection;
+            if (!this.connectionPool.TryTake(out connection, millisecondsTimeout: MillisecondsWaitingToGetConnection))
+            {
+                connection = this.connectionFactory.OpenNewConnection(this.databasePath);
+            }
+
+            return new GVFSConnection(this, connection);
+        }
+
+        private void ReturnToPool(IDbConnection connection)
+        {
+            if (this.connectionPool.IsAddingCompleted)
+            {
+                connection.Dispose();
+                return;
+            }
+
+            bool itemWasAdded = false;
+            try
+            {
+                itemWasAdded = this.connectionPool.TryAdd(connection);
+            }
+            catch (InvalidOperationException)
+            {
+                itemWasAdded = false;
+            }
+
+            if (!itemWasAdded)
+            {
+                connection.Dispose();
+            }
+        }
+
+        private void Initialize()
+        {
+            IGVFSConnectionPool connectionPool = this;

```

Comment:
Because the interface implementation is explicit
`IDbConnection IGVFSConnectionPool.GetConnection()`
You have to use the interface to call the method.

I guess we could have a private method `InternalGetConnection` or `GetConnectionImplementation`? and have the explicit interface call that method.  I can do that if it would make more sense.

Reply:
> I can do that if it would make more sense.

I think the current code is good, I missed that the implementation was explicit, thanks!